File: tool.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');
const { randomUUID } = require('node:crypto');
const { execFileSync } = require('node:child_process');
const { parseArgs } = require('node:util');
const patch = require('./package-patch');
const VERSION = '1.1.0';
const FILES = { upk: 'BrgGame/CookedPCConsole/BrgGame.upk', exe: 'Binaries/Win64/BrgGame-Steam.exe' };
const keys = Object.keys(FILES);
function gameRunning() {
  // Fail closed if process enumeration is unavailable. Never terminate a game.
  const output = execFileSync('tasklist.exe', ['/FO', 'CSV', '/NH'], { windowsHide: true, encoding: 'utf8' });
  return output.toLowerCase().includes('brggame-steam.exe');
}
const readPair = game => Object.fromEntries(keys.map(k => [k, fs.readFileSync(path.join(game, FILES[k]))]));
const hashes = pair => Object.fromEntries(keys.map(k => [k, patch.sha(pair[k])]));
const sameHashes = (a, b) => keys.every(k => a?.[k] === b?.[k]);
function buildPair(pair) {
  const upk = patch.build(pair.upk);
  return { upk, exe: patch.linkExecutable(pair.exe, pair.upk, upk) };
}
function writeExclusive(file, data) {
  const fd = fs.openSync(file, 'wx');
  try { fs.writeFileSync(fd, data); fs.fsyncSync(fd); }
  finally { fs.closeSync(fd); }
}
function saveRecord(folder, record) {
  const pending = path.join(folder, `record-${randomUUID()}.pending`);
  writeExclusive(pending, JSON.stringify(record, null, 2));
  fs.renameSync(pending, path.join(folder, 'record.json'));
}
function replacePair(game, before, after, { running = gameRunning, replace = fs.renameSync } = {}) {
  if (running()) throw new Error('게임을 완전히 종료하세요.');
  const beforeHashes = hashes(before), afterHashes = hashes(after);
  if (!sameHashes(hashes(readPair(game)), beforeHashes)) throw new Error('작업 중 게임 파일이 변경됐습니다.');
  const suffix = `.m2g-${randomUUID()}.tmp`, staged = {}, installed = [];
  try {
    for (const key of keys) {
      const temporary = path.join(game, FILES[key]) + suffix;
      // Register only after exclusive open succeeds: never delete an existing file.
      const fd = fs.openSync(temporary, 'wx'); staged[key] = temporary;
      try { fs.writeFileSync(fd, after[key]); fs.fsyncSync(fd); }
      finally { fs.closeSync(fd); }
      if (patch.sha(fs.readFileSync(temporary)) !== afterHashes[key]) throw new Error('임시 파일 검증 실패');
    }
    if (running() || !sameHashes(hashes(readPair(game)), beforeHashes)) throw new Error('설치 직전 게임 실행/파일 변경 감지');
    for (const key of keys) {
      replace(staged[key], path.join(game, FILES[key])); installed.push(key);
    }
    if (!sameHashes(hashes(readPair(game)), afterHashes)) throw new Error('설치 후 검증 실패');
  } catch (error) {
    for (const key of installed.reverse()) {
      const destination = path.join(game, FILES[key]);
      if (patch.sha(fs.readFileSync(destination)) !== afterHashes[key]) throw new Error('외부 파일 변경: 자동 복원 중단. 전용 백업을 보존했습니다.', { cause: error });
      const rollback = destination + suffix + '.rollback';
      writeExclusive(rollback, before[key]); fs.renameSync(rollback, destination);
    }
    throw error;
  } finally {
    for (const file of Object.values(staged)) if (fs.existsSync(file)) fs.unlinkSync(file);
  }
}
function apply(game, backupRoot, { running = gameRunning, builder = buildPair } = {}) {
  if (running()) throw new Error('게임을 완전히 종료하세요.');
  const before = readPair(game), after = builder(before);
  const now = new Date(), pad = n => String(n).padStart(2, '0');
  // Same sortable local-time prefix as v1.0.0 Python backups.
  const stamp = `${now.getFullYear()}${pad(now.getMonth() + 1)}${pad(now.getDate())}-${pad(now.getHours())}${pad(now.getMinutes())}${pad(now.getSeconds())}`;
  const folder = path.join(backupRoot, `${stamp}-${BigInt(Date.now()) * 1000000n}-${randomUUID().slice(0, 8)}`);
  fs.mkdirSync(backupRoot, { recursive: true }); fs.mkdirSync(folder);
  for (const key of keys) {
    const file = path.join(folder, key + '.bak');
    writeExclusive(file, before[key]);
    if (patch.sha(fs.readFileSync(file)) !== patch.sha(before[key])) throw new Error('백업 검증 실패');
  }
  const record = { version: 1, game: path.resolve(game), state: 'prepared', before: hashes(before), after: hashes(after) };
  saveRecord(folder, record);
  replacePair(game, before, after, { running });
  record.state = 'applied'; saveRecord(folder, record);
  return folder;
}
const canonical = value => path.resolve(value).toLowerCase();
function records(game, backupRoot) {
  if (!fs.existsSync(backupRoot)) return [];
  const result = [];
  for (const item of fs.readdirSync(backupRoot, { withFileTypes: true }).filter(d => d.isDirectory()).sort((a, b) => b.name.localeCompare(a.name))) {
    const folder = path.join(backupRoot, item.name), file = path.join(folder, 'record.json');
    if (!fs.existsSync(file)) continue;
    const record = JSON.parse(fs.readFileSync(file, 'utf8').replace(/^\uFEFF/, ''));
    if (record.version === 1 && typeof record.game === 'string' && canonical(record.game) === canonical(game)) {
      if (!['prepared', 'applied', 'restored'].includes(record.state) || !keys.every(k => /^[a-f0-9]{64}$/.test(record.before?.[k]) && /^[a-f0-9]{64}$/.test(record.after?.[k]))) throw new Error(`백업 기록 형식 오류: ${file}`);
      result.push({ folder, record });
    }
  }
  return result;
}
function restore(game, backupRoot, { running = gameRunning } = {}) {
  if (running()) throw new Error('게임을 완전히 종료하세요.');
  const current = readPair(game), currentHashes = hashes(current);
  for (const { folder, record } of records(game, backupRoot)) {
    if (record.state === 'restored') continue;
    if (!keys.every(k => [record.before[k], record.after[k]].includes(currentHashes[k]))) throw new Error('패치 이후 다른 도구/업데이트가 파일을 변경했습니다. 전체 백업으로 덮어쓰지 않고 중단합니다.');
    const original = Object.fromEntries(keys.map(k => [k, fs.readFileSync(path.join(folder, k + '.bak'))]));
    if (!sameHashes(hashes(original), record.before)) throw new Error('백업이 손상됐습니다.');
    replacePair(game, current, original, { running });
    record.state = 'restored'; saveRecord(folder, record);
    return folder;
  }
  throw new Error('이 게임 경로의 복원 가능한 전용 백업이 없습니다. 기존 버전의 backups 폴더를 함께 옮기세요.');
}
function detectGame() {
  const steamRoots = [path.join(process.env['ProgramFiles(x86)'] || 'C:/Program Files (x86)', 'Steam')];
  try {
    const registry = execFileSync('reg.exe', ['query', 'HKCU\\Software\\Valve\\Steam', '/v', 'SteamPath'], { windowsHide: true, encoding: 'utf8', stdio: ['ignore', 'pipe', 'pipe'] });
    const match = registry.match(/SteamPath\s+REG_SZ\s+(.+)/i);
    if (match) steamRoots.unshift(match[1].trim());
  } catch { /* Manual path and default Steam location remain available. */ }
  const roots = [...steamRoots];
  for (const steam of steamRoots) {
    const file = path.join(steam, 'steamapps/libraryfolders.vdf');
    if (fs.existsSync(file)) for (const match of fs.readFileSync(file, 'utf8').matchAll(/"path"\s+"([^"]+)"/g)) roots.push(match[1].replace(/\\\\/g, '\\'));
  }
  for (const root of new Set(roots)) {
    const game = path.join(root, 'steamapps/common/LET IT DIE');
    if (keys.every(k => fs.existsSync(path.join(game, FILES[k])))) return path.resolve(game);
  }
  return null;
}
async function main(argv = process.argv.slice(2)) {
  const { values, positionals } = parseArgs({ args: argv, allowPositionals: true, options: {
    version: { type: 'boolean' }, help: { type: 'boolean' }, yes: { type: 'boolean' }, game: { type: 'string' }, output: { type: 'string' },
  } });
  if (values.version) { console.log(VERSION); return; }
  if (values.help) { console.log('node tool.js [status|trial|apply|restore] [--game "게임 설치 폴더"] [--yes]\ntrial: --output "새 출력 폴더" 필수'); return; }
  if (positionals.length > 1 || (positionals.length && !['status', 'trial', 'apply', 'restore'].includes(positionals[0]))) throw new Error('지원하지 않는 명령입니다. --help로 확인하세요.');
  let input, lines;
  async function ask(prompt) {
    if (!input) { input = require('node:readline').createInterface({ input: process.stdin, crlfDelay: Infinity }); lines = input[Symbol.asyncIterator](); }
    process.stdout.write(prompt);
    const line = await lines.next();
    if (line.done) throw new Error('입력이 종료됐습니다.');
    return line.value.trim();
  }
  try {
    let game = values.game || detectGame(), command = positionals[0];
    if (!game && !command) game = (await ask('LET IT DIE 설치 폴더: ')).replace(/^"|"$/g, '');
    if (!game) throw new Error('--game "설치 폴더"를 지정하세요.');
    game = path.resolve(game);
    const backupRoot = path.join(__dirname, 'backups');
    if (!command) {
      console.log(`\nM2G 나이프 전용 모드 ${VERSION} · Node.js\n게임: ${game}`);
      console.log('플레이어 일반 사격만 나이프로 변경. 레이지·AI·피해 배율·비용 유지.');
      console.log('조준 룰렛 그림은 그대로지만 실제 사격은 나이프입니다.');
      console.log('다른 UPK 패치 도구는 이 모드를 먼저 복원한 뒤 사용하세요.');
      console.log('1. 상태 확인\n2. 적용\n3. 전용 백업으로 복원\n0. 종료');
      const choice = await ask('선택: ');
      if (choice === '0') return;
      command = { 1: 'status', 2: 'apply', 3: 'restore' }[choice];
      if (!command) throw new Error('잘못된 선택');
    }
    if (command === 'status') {
      const pair = readPair(game), current = hashes(pair);
      if (records(game, backupRoot).some(({ record }) => record.state !== 'restored' && sameHashes(record.after, current))) console.log('나이프 전용 적용됨. 기존 버전 백업도 지원합니다.');
      else { buildPair(pair); console.log('지원되는 M2G 함수/실행 파일 해시 연결 확인. 현재 미적용.'); }
    } else if (command === 'trial') {
      if (!values.output) throw new Error('trial은 --output "새 출력 폴더"가 필요합니다.');
      const before = readPair(game), after = buildPair(before), output = path.resolve(values.output);
      fs.mkdirSync(output);
      for (const key of keys) writeExclusive(path.join(output, path.basename(FILES[key])), after[key]);
      if (!sameHashes(hashes(readPair(game)), hashes(before))) throw new Error('검증 중 원본이 변경됐습니다.');
      console.log(`복사본 생성 완료: ${output}\n원본 게임은 변경하지 않았습니다.`);
    } else {
      if (!values.yes && !['y', 'yes'].includes((await ask(`게임 종료 후 실행하세요. ${command} 진행? (y/N): `)).toLowerCase())) return;
      console.log('파일 검증 및 처리 중입니다. 창을 닫지 마세요.');
      const folder = command === 'apply' ? apply(game, backupRoot) : restore(game, backupRoot);
      console.log(`${command} 완료. 백업: ${folder}`);
    }
  } finally { input?.close(); }
}
if (require.main === module) main().catch(error => {
  console.error(`오류: ${error.message}\n파일 접근이 거부되면 터미널을 관리자 권한으로 실행하세요.`);
  process.exitCode = 1;
});
module.exports = { VERSION, FILES, gameRunning, readPair, hashes, buildPair, saveRecord, replacePair, apply, records, restore, detectGame, main };

File: tool.py
"""LET IT DIE M2G player knife-only mode; local offline game only."""
import argparse
import json
import os
from pathlib import Path
import re
import subprocess
import time
import uuid

from package_patch import build, link_executable, sha

ROOT = Path(__file__).resolve().parent
VERSION = '1.0.0'
FILES = {'upk': 'BrgGame/CookedPCConsole/BrgGame.upk', 'exe': 'Binaries/Win64/BrgGame-Steam.exe'}


def game_running():
    result = subprocess.run(['tasklist.exe', '/FO', 'CSV', '/NH'], capture_output=True, check=True)
    return b'brggame-steam.exe' in result.stdout.lower()


def read_pair(game):
    return {key: (game / relative).read_bytes() for key, relative in FILES.items()}


def hashes(pair):
    return {key: sha(data) for key, data in pair.items()}


def build_pair(pair):
    upk = build(pair['upk'])
    return {'upk': upk, 'exe': link_executable(pair['exe'], pair['upk'], upk)}


def save_record(folder, record):
    pending = folder / ('record-' + uuid.uuid4().hex + '.pending')
    with pending.open('x', encoding='utf-8') as stream:
        json.dump(record, stream, ensure_ascii=False, indent=2)
    pending.replace(folder / 'record.json')


def replace_pair(game, before, after, replace=os.replace, running=game_running):
    if running():
        raise ValueError('게임을 완전히 종료하세요.')
    if hashes(read_pair(game)) != hashes(before):
        raise ValueError('작업 중 게임 파일이 변경됐습니다.')
    suffix = '.m2g-' + uuid.uuid4().hex + '.tmp'
    staged = {}
    installed = []
    try:
        for key, relative in FILES.items():
            temporary = Path(str(game / relative) + suffix)
            with temporary.open('xb') as stream:
                stream.write(after[key])
                stream.flush()
                os.fsync(stream.fileno())
            staged[key] = temporary
            if sha(temporary.read_bytes()) != sha(after[key]):
                raise ValueError('임시 파일 검증 실패')
        if running() or hashes(read_pair(game)) != hashes(before):
            raise ValueError('설치 직전 게임 실행/파일 변경 감지')
        for key, relative in FILES.items():
            replace(staged[key], game / relative)
            installed.append(key)
        if hashes(read_pair(game)) != hashes(after):
            raise ValueError('설치 후 검증 실패')
    except Exception:
        # Roll back only files replaced by this transaction. Unknown externally
        # modified data is never silently overwritten.
        for key in reversed(installed):
            destination = game / FILES[key]
            if sha(destination.read_bytes()) != sha(after[key]):
                raise RuntimeError('외부 파일 변경: 자동 복원 중단. 전용 백업을 보존했습니다.')
            rollback = Path(str(destination) + suffix + '.rollback')
            with rollback.open('xb') as stream:
                stream.write(before[key])
            os.replace(rollback, destination)
        raise
    finally:
        for temporary in staged.values():
            if temporary.exists():
                temporary.unlink()


def apply(game, backup_root, running=game_running):
    if running():
        raise ValueError('게임을 완전히 종료하세요.')
    before = read_pair(game)
    after = build_pair(before)
    folder = backup_root / (time.strftime('%Y%m%d-%H%M%S-') + str(time.time_ns()) + '-' + uuid.uuid4().hex[:8])
    folder.mkdir(parents=True, exist_ok=False)
    for key, data in before.items():
        with (folder / (key + '.bak')).open('xb') as stream:
            stream.write(data)
        if sha((folder / (key + '.bak')).read_bytes()) != sha(data):
            raise ValueError('백업 검증 실패')
    record = {'version': 1, 'game': str(game.resolve()), 'state': 'prepared', 'before': hashes(before), 'after': hashes(after)}
    save_record(folder, record)
    replace_pair(game, before, after, running=running)
    record['state'] = 'applied'
    save_record(folder, record)
    return folder


def records(game, backup_root):
    result = []
    for file in sorted(backup_root.glob('*/record.json'), reverse=True):
        record = json.loads(file.read_text(encoding='utf-8'))
        if record.get('version') == 1 and os.path.normcase(record.get('game', '')) == os.path.normcase(str(game.resolve())):
            result.append((file.parent, record))
    return result


def restore(game, backup_root, running=game_running):
    if running():
        raise ValueError('게임을 완전히 종료하세요.')
    current = read_pair(game)
    candidates = records(game, backup_root)
    for folder, record in candidates:
        if record['state'] == 'restored':
            continue
        # A prepared record can recover an interrupted two-file installation.
        current_hashes = hashes(current)
        if not all(current_hashes[k] in (record['before'][k], record['after'][k]) for k in FILES):
            raise ValueError('패치 이후 다른 도구/업데이트가 파일을 변경했습니다. 전체 백업으로 덮어쓰지 않고 중단합니다.')
        original = {key: (folder / (key + '.bak')).read_bytes() for key in FILES}
        if hashes(original) != record['before']:
            raise ValueError('백업이 손상됐습니다.')
        replace_pair(game, current, original, running=running)
        record['state'] = 'restored'
        save_record(folder, record)
        return folder
    raise ValueError('이 게임 경로의 복원 가능한 전용 백업이 없습니다.')


def detect_game():
    steam_roots = [Path(os.environ.get('ProgramFiles(x86)', 'C:/Program Files (x86)')) / 'Steam']
    try:
        import winreg
        with winreg.OpenKey(winreg.HKEY_CURRENT_USER, r'Software\Valve\Steam') as key:
            steam_roots.insert(0, Path(winreg.QueryValueEx(key, 'SteamPath')[0]))
    except (ImportError, OSError):
        pass
    roots = list(steam_roots)
    for steam in steam_roots:
        library = steam / 'steamapps/libraryfolders.vdf'
        if library.exists():
            roots.extend(Path(x.replace('\\\\', '\\')) for x in re.findall(r'"path"\s+"([^"]+)"', library.read_text(encoding='utf-8')))
    for root in roots:
        game = root / 'steamapps/common/LET IT DIE'
        if all((game / relative).is_file() for relative in FILES.values()):
            return game.resolve()
    return None


def main():
    parser = argparse.ArgumentParser(description=f'M2G 플레이어 나이프 전용 모드 {VERSION} 정식 버전')
    parser.add_argument('--version', action='version', version=VERSION)
    parser.add_argument('command', nargs='?', choices=['status', 'trial', 'apply', 'restore'])
    parser.add_argument('--game', type=Path)
    parser.add_argument('--output', type=Path)
    parser.add_argument('--yes', action='store_true')
    args = parser.parse_args()
    game = args.game or detect_game()
    if game is None and args.command is None:
        game = Path(input('LET IT DIE 설치 폴더: ').strip().strip('"'))
    if game is None:
        parser.error('--game 설치폴더를 지정하세요.')
    game = game.resolve()
    command = args.command
    backup_root = ROOT / 'backups'
    if command is None:
        print(f'\nM2G 나이프 전용 모드 {VERSION}\n게임: {game}')
        print('플레이어 일반 사격만 나이프로 변경. 레이지·AI·피해 배율·비용 유지.')
        print('조준 룰렛 그림은 그대로지만 실제 사격은 나이프입니다. 사용자 실게임 정상 작동 확인.')
        print('다른 UPK 패치 도구는 이 모드를 먼저 복원한 뒤 사용하세요.')
        print('1. 상태 확인\n2. 적용\n3. 전용 백업으로 복원\n0. 종료')
        choice = input('선택: ').strip()
        if choice == '0':
            return
        command = {'1': 'status', '2': 'apply', '3': 'restore'}.get(choice)
        if command is None:
            raise ValueError('잘못된 선택')
    if command == 'status':
        current = hashes(read_pair(game))
        active = any(record['after'] == current for _, record in records(game, backup_root))
        if active:
            print('나이프 전용 적용됨. 사용자 실게임 정상 작동 확인 버전입니다.')
        else:
            build_pair(read_pair(game))
            print('지원되는 M2G 함수/실행 파일 해시 연결 확인. 현재 미적용.')
    elif command == 'trial':
        if args.output is None:
            parser.error('trial은 --output 새 출력폴더가 필요합니다.')
        before = read_pair(game)
        after = build_pair(before)
        output = args.output.resolve()
        output.mkdir(parents=True, exist_ok=False)
        for key, data in after.items():
            (output / Path(FILES[key]).name).write_bytes(data)
        if hashes(read_pair(game)) != hashes(before):
            raise ValueError('검증 중 원본이 변경됐습니다.')
        print(f'복사본 생성 완료: {output}\n원본 게임은 변경하지 않았습니다.')
    else:
        if not args.yes and input('게임 종료 후 실행하세요. '+command+' 진행? (y/N): ').lower() not in ('y', 'yes'):
            return
        folder = apply(game, backup_root) if command == 'apply' else restore(game, backup_root)
        print(f'{command} 완료. 백업: {folder}')


if __name__ == '__main__':
    try:
        main()
    except Exception as error:
        print(f'오류: {error}\n파일 접근이 거부되면 터미널을 관리자 권한으로 실행하세요.')
        raise SystemExit(1)
